docs/source/model_library.rst

```diff
@@ -0,0 +1,550 @@
+the ``some_other_model`` wasn't borrowed from the library). To replace
+a model in the library you will need to first borrow the model at the index
+you want to use and provide the index to the call to
+`~stpipe.library.AbstractModelLibrary.shelve`.
+
+.. code-block:: pycon
+
+   >>> with library:
+   ...     library.borrow(0)
+   ...     library.shelve(some_other_model, 0)
+
+Forgetting to first borrow the model at the index will also produce a
+`~stpipe.library.BorrowError` (even if you provide the index).
+
+.. code-block:: pycon
+
+   >>> with library:
+   ...     library.shelve(some_other_model, 0)
+
+   BorrowError: Attempt to shelve model at a non-borrowed index
+
+.. _library_implementing_a_subclass:
+
+Implementing a subclass
+=======================
+
+Several methods are abstract and will need implementations:
+
+- Methods used by stpipe:
+
+  - `~stpipe.library.AbstractModelLibrary.crds_observatory`
+
+- Methods used by `~stpipe.library.AbstractModelLibrary`
+
+  - ``_datamodels_open``
+  - ``_load_asn``
+  - ``_filename_to_group_id``
+  - ``_model_to_group_id``
+
+It's likely that a few other methods might require overriding:
+
+- ``_model_to_filename``
+- ``_assign_member_to_model``
+
+Consult the docstrings (and base implementations) for more details.
+
+It may also be required (depending on your usage) to update
+``stpipe.step.Step._datamodels_open`` to allow stpipe to open and inspect an
+`~stpipe.library.AbstractModelLibrary` when provided as a `~stpipe.step.Step` input.
+
+.. _library_developer_documentation:
+
+Developer Documentation
+=======================
+
+What follows are note primarily aimed towards developers and
+maintainers of `~stpipe.library.AbstractModelLibrary`. This section might be useful
+to provide context to users but shouldn't be necessary for a user
+to effectively use `~stpipe.library.AbstractModelLibrary`.
+
+.. _library_motivation:
+
+Motivation
+----------
+
+The development of `~stpipe.library.AbstractModelLibrary` was largely motivated by
+the need for a container compatible with stpipe machinery
+that would allow passing "on disk" models between steps. Existing
+containers (when used in "memory saving" modes) were not compatible
+with stpipe. These containers also sometimes allowed input files
+to be overwritten. It was decided that a new container would be
+developed to address these and other issues. This would allow
+gradual migration for pipeline code where specific steps and pipelines
+could update to `~stpipe.library.AbstractModelLibrary` while leaving the existing
+container unchanged for other steps.
+
+A survey of container usage was performed with a few key findings:
+
+- Many uses could be replaced by simpler containers (lists)
+- When loaded from an association, the container size never changed
+- The order of models was never changed
+- Needs various methods for stpipe
+- Several steps implemented different memory optimizations
```

Comment:
this feels like two bullet points

Reply:
Good point. I added this as one because the complexity was due to the memory optimization (if the container always returned either a filename or a model the step code would be simpler). Does the split in f350f55 work?

Reply:
yeah i think this is fine